### PR TITLE
Improve nodeos make-index speeds

### PR DESCRIPTION
**Pull request by brianjohnson5972**
## Change Description
Changed nodeos to be more efficient in re-creating an index file.  Code is a refactoring of original eosio-blocklog implementation for make-index and both implementations use it.

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: libraries/chain/block_log.cpp [hunk 1]
```diff
@@ -317,44 +376,37 @@ namespace eosio { namespace chain {
...
-      if (my->version == 1) {
-         pos = 4; // Skip version which should have already been checked.
-      } else {
-         pos = 8; // Skip version and first block offset which should have already been checked
+   void block_log::construct_index(const fc::path& block_file_name, const fc::path& index_file_name, bool provide_index_log_status) {
+      detail::reverse_iterator block_log_iter;
+
+      const uint32_t num_blocks = block_log_iter.open(block_file_name);
+      if (provide_index_log_status) {
+         std::cout << "\nblock log version= " << block_log_iter.version() << '\n';
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: I was trying to keep the logging the same as it was, so not present for nodeos and no extra context for presenting in eosio-blocklog

[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/block_log.cpp]: I see no reason to maintain existing logging output.

File: libraries/chain/block_log.cpp [hunk 2]
```diff
@@ -68,6 +70,63 @@ namespace eosio { namespace chain {
...
+
+         FILE*                              _file              = nullptr;
+         const std::string                  _block_index_name;
+         const uint32_t                     _blocks_expected;
+         uint32_t                           _block_written;
+         bool                               _report_status;
+         std::unique_ptr<uint64_t[]>        _buffer_ptr;
+         int64_t                            _current_position;
+         int64_t                            _start_of_buffer_position;
+         int64_t                            _end_of_buffer_position;
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: ok

File: libraries/chain/block_log.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: ok

File: libraries/chain/block_log.cpp [hunk 3]
```diff
@@ -317,44 +376,37 @@ namespace eosio { namespace chain {
...
+      if (num_blocks == 0) {
+         return;
+      }
 
-      // skip the totem
-      if (my->version > 1) {
-         uint64_t totem;
-         my->block_stream.read((char*) &totem, sizeof(totem));
+      if (provide_index_log_status) {
+         std::cout << "\nfirst block= " << block_log_iter.first_block_num() << ", last block= " << (block_log_iter.first_block_num() + num_blocks) << "\n\n";
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: ok

File: libraries/chain/block_log.cpp [hunk 4]
```diff
@@ -543,4 +595,201 @@ namespace eosio { namespace chain {
...
    }
 
+   detail::reverse_iterator::reverse_iterator()
+   : _buffer_ptr(std::make_unique<char[]>(_buf_len)) {
+   }
+
+   detail::reverse_iterator::~reverse_iterator() {
+      if (_file != nullptr) {
+         fclose(_file);
+      }
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: ok

File: libraries/chain/block_log.cpp [hunk 5]
```diff
@@ -5,10 +5,12 @@
 #include <eosio/chain/block_log.hpp>
 #include <eosio/chain/exceptions.hpp>
 #include <fstream>
+#include <fc/bitutil.hpp>
 #include <fc/io/raw.hpp>
 
 #define LOG_READ  (std::ios::in | std::ios::binary)
 #define LOG_WRITE (std::ios::out | std::ios::binary | std::ios::app)
+#define LOG_WRITE_NEW (std::ios::out | std::ios::binary)
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: ok

File: libraries/chain/block_log.cpp [hunk 6]
```diff
@@ -543,4 +595,201 @@ namespace eosio { namespace chain {
...
+      }
+   }
+
+   void detail::index_writer::write(uint64_t pos) {
+      prepare_buffer();
+      uint64_t* buffer = _buffer_ptr.get();
+      buffer[_current_position - _start_of_buffer_position] = pos;
+      --_current_position;
+      if (_report_status && (_block_written & 0xfffff) == 0) {                            //periodically print a progress indicator
+         std::cout << "block: " << std::setw(10) << _block_written << "    position in file " << std::setw(14) << pos << '\n';
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: ok

File: libraries/chain/block_log.cpp [hunk 7]
```diff
@@ -543,4 +595,201 @@ namespace eosio { namespace chain {
...
+      _end_of_buffer_position = _current_position_in_file + _position_size;
+      if (_end_of_buffer_position < _buf_len) {
+         _start_of_buffer_position = 0;
+      }
+      else {
+         _start_of_buffer_position = _end_of_buffer_position - _buf_len;
+      }
+
+      auto status = fseek(_file, _start_of_buffer_position, SEEK_SET);
+      EOS_ASSERT( status == 0, block_log_exception, "Could not seek in '${blocks_log}' to position: ${pos}", ("blocks_log", _block_file_name)("pos", _start_of_buffer_position) );
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: ok

File: libraries/chain/block_log.cpp [hunk 8]
```diff
@@ -543,4 +595,201 @@ namespace eosio { namespace chain {
...
+
+   bool detail::index_writer::shift_buffer() {
+      if (_current_position >= _start_of_buffer_position) {
+         return false;
+      }
+
+      const auto file_location_start = buffer_location_to_file_location(_start_of_buffer_position);
+
+      auto status = fseek(_file, file_location_start, SEEK_SET);
+      EOS_ASSERT( status == 0, block_log_exception, "Could not navigate in '${blocks_index}' file_location_start: ${loc}, _start_of_buffer_position: ${_start_of_buffer_position}", ("blocks_index", _block_index_name)("loc", file_location_start)("_start_of_buffer_position",_start_of_buffer_position) );
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: ok

File: libraries/chain/block_log.cpp [hunk 9]
```diff
@@ -543,4 +595,201 @@ namespace eosio { namespace chain {
...
+      EOS_ASSERT( _eof_position_in_file > 0, block_log_exception, "Block log file at '${blocks_log}' could not be read.", ("blocks_log", _block_file_name) );
+      _current_position_in_file = _eof_position_in_file - _position_size;
+
+      update_buffer();
+
+      _blocks_found = 0;
+      char* buf = _buffer_ptr.get();
+      const uint32_t index_of_pos = _current_position_in_file - _start_of_buffer_position;
+      const uint64_t block_pos = *reinterpret_cast<uint64_t*>(buf + index_of_pos);
+      uint32_t bnum;
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/block_log.cpp]: ok

**Pull request merged**